javascript/ql/src/Security/CWE-078/IndirectCommandInjection.ql

```diff
@@ -0,0 +1,38 @@
+/**
+ * @name Indirect uncontrolled command line
+ * @description Using command line parameters to construct other
+ *              command line arguments may allow a malicious user to
+ *              indirectly change the meaning of the command.
+ * @kind path-problem
+ * @problem.severity warning
+ * @precision medium
+ * @id js/indirect-command-line-injection
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-078
+ *       external/cwe/cwe-088
+ */
+
+import javascript
+import DataFlow::PathGraph
+import semmle.javascript.security.dataflow.IndirectCommandInjection::IndirectCommandInjection
+
+/**
+ * Holds if `result` is `src` or a slice thereof.
+ */
+DataFlow::SourceNode getACommandLineArgumentsArray(CommandLineArgumentsArray src) {
+  src = result
+  or
+  getACommandLineArgumentsArray(src).getAMethodCall("slice") = result
+}
+
+from Configuration cfg, DataFlow::PathNode source, DataFlow::PathNode sink, DataFlow::Node highlight
+where
+  cfg.hasFlowPath(source, sink) and
+  // avoid flagging trivial CLI utilities by excluding simple, direct flow from command line arguments to sinks
+  not getACommandLineArgumentsArray(source.getNode()).getAPropertyRead().flowsTo(sink.getNode()) and
```

Comment:
Am I right in thinking that this is simply a heuristic? There isn't anything that makes single parameters intrinsically more resilient to unexpected substitutions than multiple parameters, is there?

Reply:
Yes, it is just a heuristic. But with the reduced precision, and detailed qhelp, I think we can remove that whitelist. I have started another evaluation for just the removal of the whitelist.

Reply:
`process.argv[0]` and `process.argv[1]` are paths to `node` and `main`, they are the only ones we need to whitelist according to default.slugs. I have done that with a simple `Sanitizer` instance.